django/contrib/admin/options.py

```diff
@@ -1201,15 +1202,39 @@ def construct_search(field_name):
                     if hasattr(field, "path_infos"):
                         # Update opts to follow the relation.
                         opts = field.path_infos[-1].to_opts
+            # Return the field name as-is if it already includes __exact.
+            if field_name.endswith("__exact"):
+                return field_name
             # Otherwise, use the field with icontains.
             return "%s__icontains" % field_name
 
         may_have_duplicates = False
         search_fields = self.get_search_fields(request)
         if search_fields and search_term:
-            orm_lookups = [
-                construct_search(str(search_field)) for search_field in search_fields
-            ]
+            # Add string cast annotations for non-string exact lookups.
+            str_annotations = {}
+            orm_lookups = []
+            for field in search_fields:
+                if field.endswith("__exact"):
+                    field_name = field.rsplit("__exact", 1)[0]
+                    try:
+                        field_obj = queryset.model._meta.get_field(field_name)
+                    except FieldDoesNotExist:
+                        lookup = construct_search(field)
+                        orm_lookups.append(lookup)
+                        continue
+                    if not isinstance(field_obj, (models.CharField, models.TextField)):
+                        str_annotations[f"{field_name}_str"] = Cast(
+                            field_name, output_field=models.CharField()
+                        )
+                        lookup = construct_search(f"{field_name}_str__exact")
+                        orm_lookups.append(lookup)
+                    else:
+                        lookup = construct_search(field)
+                        orm_lookups.append(lookup)
+                else:
+                    lookup = construct_search(str(field))
+                    orm_lookups.append(lookup)
```

Comment:
```suggestion
                    orm_lookups.append(lookup)

            if str_annotations:
                queryset = queryset.annotate(**str_annotations)
```
(sorry I forgot this)